[wip] fast typeMeta/ScalarType conversion approach 1

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45024 pass TypeMeta by value
* **#44456 [wip] fast TypeMetaData instances for ScalarType types**

This PR implements a simple fast mapping between TypeMeta values and ScalarType types, as described below. However, #44965 is compellingly better in both speed and size, so that stack supersedes this one. Leaving it open for comparison purposes for now.

---
[Old notes]

(Re)introduce a fast mapping between TypeMeta values representing ScalarType types, and the corresponding ScalarType enum values. Fast conversion between the two representations is important due to the number of times such conversions occur currently. 

TypeMetaData instances are distinguished by unique type ids. In a simpler bygone world, type ids were passed to the TypeMetaData constructor explicitly, and a fast mapping was implemented simply by "preallocating" the ScalarType enums as type ids for the relevant TypeMetaData instances. 

However this scheme was subsequently replaced by one which uses [hashes of compile-time type names for type ids](https://github.com/pytorch/pytorch/blob/master/c10/util/TypeIndex.h#L57), and there are signs that [this relation is used elsewhere in the system](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/ivalue.cpp#L698). So rather than specifying type ids by fiat, this PR simply adds an additional, optional data member to TypeMetaData which holds the corresponding ScalarType enum value, if applicable. 

Perf improvement looks good  - up to ~10% speedup in FB microbenchmarks (experiment is reachable from the diff below) with plausible corresponding Ir reduction in callgrind. 

Other notes: 
* there's less to this PR than meets the eye. The nontrivial changes are in `c10/util/typeid.h` and `c10/util/typeid.cpp`; the rest of the diffs are just code movement: TypeMeta<->ScalarType conversion functions have been moved to the new `c10/core/ScalarTypeToTypeMeta.h` (and lightly adapted to take advantage of the fast mapping). This is so that typeid.h can depend on ScalarType.h without creating a circular dependency. Given the opportunity, I went ahead and jittered some #includes to clarify when the conversion functions (rather than just ScalarType) were being used.
* adding a second constructor to TypeMetaData caused clang to silently treat neither constructor as `constexpr`, despite both being annotated as such (and the original being unchanged). Rather than work around this, I just removed the downstream constexpr annotations which required the constructor to be constexpr (the optimization scheme that relied on them is incomplete anyway - e.g., it's disabled in GCC - and isn't needed by the setup introduced in this PR)

Differential Revision: [D23626808](https://our.internmc.facebook.com/intern/diff/D23626808)

Commit:
eager static init for TypeMetaData instances

[ghstack-poisoned]

Commit:
Update on "eager static init for TypeMetaData instances"


Differential Revision: [D23626808](https://our.internmc.facebook.com/intern/diff/D23626808)

[ghstack-poisoned]

Commit:
Update on "fast TypeMetaData instances for ScalarType types"


Differential Revision: [D23626808](https://our.internmc.facebook.com/intern/diff/D23626808)

[ghstack-poisoned]

Commit:
Update on "fast TypeMetaData instances for ScalarType types"


Differential Revision: [D23626808](https://our.internmc.facebook.com/intern/diff/D23626808)

[ghstack-poisoned]

Commit:
Update on "fast TypeMetaData instances for ScalarType types"


Differential Revision: [D23626808](https://our.internmc.facebook.com/intern/diff/D23626808)

[ghstack-poisoned]

Commit:
Update on "[wip] fast TypeMetaData instances for ScalarType types"


[wip, wanna try something]

Here we reintroduce a fast mapping between the particular TypeMetaData instances that represent C++ types enumerated by ScalarType, and the corresponding ScalarType values. This facilitates lightweight conversions between the two representations, critical due to the number of times such conversions occur currently. 

TypeMetaData instances are distinguished by unique type ids. In a simpler bygone world, type ids were passed to the TypeMetaData constructor explicitly, and a fast mapping was implemented simply by "preallocating" the ScalarType enums as type ids for the relevant TypeMetaData instances. 

However this scheme was subsequently replaced by one which uses [hashes of compile-time type names for type ids](https://github.com/pytorch/pytorch/blob/master/c10/util/TypeIndex.h#L57), and there are signs that [this relation is used elsewhere in the system](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/ivalue.cpp#L698). So rather than specifying type ids by fiat, this PR simply adds an additional, optional data member to TypeMetaData which holds the corresponding ScalarType enum value, if applicable. 

Reviewer note: there's less to this PR than meets the eye. The nontrivial changes are in `c10/util/typeid.h` and `c10/util/typeid.cpp`; the rest of the diffs are just code movement: TypeMeta<->ScalarType conversion functions have been moved to the new `c10/core/ScalarTypeToTypeMeta.h` (and lightly adapted to take advantage of the fast mapping). This is so that typeid.h can depend on ScalarType.h without creating a circular dependency. Given the opportunity, I went ahead and jittered some #includes to clarify when the conversion functions (rather than just ScalarType) were being used.

Differential Revision: [D23626808](https://our.internmc.facebook.com/intern/diff/D23626808)

[ghstack-poisoned]

ezyang commented on c10/util/typeid.h:
```diff
@@ -126,7 +123,8 @@ struct TypeMetaData final {
   TypeMetaData() = delete;
-  constexpr TypeMetaData(
+
+  TypeMetaData(
```
Why not constexpr anymore?

bhosmer replied:
See the second note in the PR description - adding the second constructor caused clang to start erroring as if the `constexpr` annotation had been silently dropped from both constructors. The client code didn't really rely on it so I just dropped them - but FYI there's no second constructor in #44965 so it's back 😁

ezyang commented on c10/util/typeid.cpp:
```diff
@@ -20,36 +20,62 @@ C10_EXPORT void _ThrowRuntimeTypeLogicError(const string& msg) {
+
+// Canonical TypeMetaData instances for C++ types enumerated by ScalarType
+namespace detail {
+static TypeMetaData scalar_type_metas[] = {
```
Top level static. Are you sure you aren't subject to SIOF here?

bhosmer replied:
Yeah, good point. Final version of this PR simplifies this array out of existence, and #44965 shouldn't have any top level statics.

ezyang commented on c10/core/ScalarTypeToTypeMeta.h:
```diff
@@ -0,0 +1,62 @@
+ * convert TypeMeta handles to ScalarType enum values
+ * Caution: extremely hot
+ */
+static inline ScalarType typeMetaToScalarType(caffe2::TypeMeta dtype) {
```
Re SIOF problems: would you bet the farm on this function never being called in a static initializer? I wouldn't...

bhosmer replied:
Yeah agree

ezyang commented:
This looks plausible but I am concerned about SIOF.

dzhulgakov commented on c10/core/ScalarTypeToTypeMeta.h:
```diff
@@ -0,0 +1,62 @@
+  case ScalarType::name:         \
+    return caffe2::TypeMeta::Make<ctype>();
+
+  switch (scalar_type) {
```
maybe instead of a `switch` do a pointer arithmetic on the array? I bet it might be even faster

bhosmer replied:
True, though the purest version of this approach didn't have the global array at all (PR updated now), which is why I wasn't depending on it here. Also, this is a pretty cold path - it's called ~0 times in the benchmarks I'm using.

But all that said, one advantage of #44965 is that it doesn't even need pointer arithmetic for this :)

dzhulgakov commented on c10/util/typeid.h:
```diff
@@ -308,6 +328,22 @@ inline C10_TYPENAME_CONSTEXPR TypeMetaData _makeTypeMetaDataInstance() {
 }
 
+template <class T>
+inline TypeMetaData _makeScalarTypeMetaDataInstance(ScalarType scalarType) {
```
why is this not constexpr anymore? if it was - we could have guaranteed that no SIOF occurs I think

bhosmer replied:
I don't think it guaranteed anything - e.g. the macro is a no-op for old GCC etc.

dzhulgakov commented on c10/util/typeid.h:
```diff
@@ -152,6 +171,7 @@ struct TypeMetaData final {
   Delete* delete_;
   TypeIdentifier id_;
   c10::string_view name_;
+  c10::optional<ScalarType> scalarTypeOpt_;
```
technically you don't need this one and could resort to just the pointer arithmetic (it's faster!)

bhosmer replied:
See #44965 😁

dzhulgakov reviewed (commented):
Awesome! One thing unclear to me - why did you have to remove a bunch of constexprs?

dzhulgakov commented:
> there are signs that this relation is used elsewhere in the system

it seems that the link is off, where exactly is it used?

Commit:
Update on "[wip] fast TypeMetaData instances for ScalarType types"

[wip, one more idea :P]


(Re)introduce a fast mapping between TypeMeta values representing ScalarType types, and the corresponding ScalarType enum values. Fast conversion between the two representations is important due to the number of times such conversions occur currently. 

TypeMetaData instances are distinguished by unique type ids. In a simpler bygone world, type ids were passed to the TypeMetaData constructor explicitly, and a fast mapping was implemented simply by "preallocating" the ScalarType enums as type ids for the relevant TypeMetaData instances. 

However this scheme was subsequently replaced by one which uses [hashes of compile-time type names for type ids](https://github.com/pytorch/pytorch/blob/master/c10/util/TypeIndex.h#L57), and there are signs that [this relation is used elsewhere in the system](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/ivalue.cpp#L698). So rather than specifying type ids by fiat, this PR simply adds an additional, optional data member to TypeMetaData which holds the corresponding ScalarType enum value, if applicable. 

Perf improvement looks good  - up to ~10% speedup in FB microbenchmarks (experiment is reachable from the diff below) with plausible corresponding Ir reduction in callgrind. 

Other notes: 
* there's less to this PR than meets the eye. The nontrivial changes are in `c10/util/typeid.h` and `c10/util/typeid.cpp`; the rest of the diffs are just code movement: TypeMeta<->ScalarType conversion functions have been moved to the new `c10/core/ScalarTypeToTypeMeta.h` (and lightly adapted to take advantage of the fast mapping). This is so that typeid.h can depend on ScalarType.h without creating a circular dependency. Given the opportunity, I went ahead and jittered some #includes to clarify when the conversion functions (rather than just ScalarType) were being used.
* adding a second constructor to TypeMetaData caused clang to silently treat neither constructor as `constexpr`, despite both being annotated as such (and the original being unchanged). Rather than work around this, I just removed the downstream constexpr annotations which required the constructor to be constexpr (the optimization scheme that relied on them is incomplete anyway - e.g., it's disabled in GCC - and isn't needed by the setup introduced in this PR)

Differential Revision: [D23626808](https://our.internmc.facebook.com/intern/diff/D23626808)

[ghstack-poisoned]

Commit:
Update on "[wip] fast TypeMetaData instances for ScalarType types"

[wip, one more idea :P]


(Re)introduce a fast mapping between TypeMeta values representing ScalarType types, and the corresponding ScalarType enum values. Fast conversion between the two representations is important due to the number of times such conversions occur currently. 

TypeMetaData instances are distinguished by unique type ids. In a simpler bygone world, type ids were passed to the TypeMetaData constructor explicitly, and a fast mapping was implemented simply by "preallocating" the ScalarType enums as type ids for the relevant TypeMetaData instances. 

However this scheme was subsequently replaced by one which uses [hashes of compile-time type names for type ids](https://github.com/pytorch/pytorch/blob/master/c10/util/TypeIndex.h#L57), and there are signs that [this relation is used elsewhere in the system](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/ivalue.cpp#L698). So rather than specifying type ids by fiat, this PR simply adds an additional, optional data member to TypeMetaData which holds the corresponding ScalarType enum value, if applicable. 

Perf improvement looks good  - up to ~10% speedup in FB microbenchmarks (experiment is reachable from the diff below) with plausible corresponding Ir reduction in callgrind. 

Other notes: 
* there's less to this PR than meets the eye. The nontrivial changes are in `c10/util/typeid.h` and `c10/util/typeid.cpp`; the rest of the diffs are just code movement: TypeMeta<->ScalarType conversion functions have been moved to the new `c10/core/ScalarTypeToTypeMeta.h` (and lightly adapted to take advantage of the fast mapping). This is so that typeid.h can depend on ScalarType.h without creating a circular dependency. Given the opportunity, I went ahead and jittered some #includes to clarify when the conversion functions (rather than just ScalarType) were being used.
* adding a second constructor to TypeMetaData caused clang to silently treat neither constructor as `constexpr`, despite both being annotated as such (and the original being unchanged). Rather than work around this, I just removed the downstream constexpr annotations which required the constructor to be constexpr (the optimization scheme that relied on them is incomplete anyway - e.g., it's disabled in GCC - and isn't needed by the setup introduced in this PR)

Differential Revision: [D23626808](https://our.internmc.facebook.com/intern/diff/D23626808)

[ghstack-poisoned]

bhosmer commented:
> Awesome! One thing unclear to me - why did you have to remove a bunch of constexprs?

See the second note in the description :) basically they weren't being relied on (e.g., in oldish versions of GCC they were deffed out) and I triggered some issue in clang when I added a second constructor to TypeMetaData. 

But in #44965 which supersedes this, the constructor's constexpr is back (though not the macro-defined ones, they didn't help anything and just obfuscated the code)

bhosmer commented:
> > there are signs that this relation is used elsewhere in the system
> 
> it seems that the link is off, where exactly is it used?

AH turns out I conflated `std::type_index` and `c10::util::type_index` - [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/ivalue.cpp#L816)'s the fixed link FWIW, but yeah maybe it's actually not used anywhere else.

bhosmer commented:
superceded by #44965
